src/app/rpmostree-builtin-kargs.cxx

```diff
@@ -291,33 +283,53 @@ rpmostree_builtin_kargs (int            argc,
         return FALSE;
 
       const char* current_kernel_arg_string = NULL;
+      gboolean kargs_changed = FALSE;
       if (!kernel_arg_handle_editor (old_kernel_arg_string, &current_kernel_arg_string,
-                                     cancellable, error))
+                                     &kargs_changed, cancellable, error))
         return FALSE;
-      gboolean out_changed = FALSE;
+      if (!kargs_changed)
+        {
+          if (opt_unchanged_exit_77)
+            {
+              g_print("No changes.\n");
+              invocation->exit_code = RPM_OSTREE_EXIT_UNCHANGED;
+              return TRUE;
+            }
+          else
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+                           "The kernel arguments remained the same");
+              return FALSE;
+            }
+        }
 
       /* Here we load the sysroot again, if the sysroot
        * has been changed since then, we directly error
        * out.
        */
+      gboolean sysroot_changed = FALSE;
       if (!ostree_sysroot_load_if_changed (before_sysroot,
-                                           &out_changed,
+                                           &sysroot_changed,
                                            cancellable,
                                            error))
         return FALSE;
-      if (out_changed)
+      if (sysroot_changed)
         {
           g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
                        "Conflict: bootloader configuration changed. Saved kernel arguments: \n%s", current_kernel_arg_string);
 
           return FALSE;
         }
-      /* We use the user defined kernel args as existing arguments here
-       * and kept other strvs empty, because the existing kernel arguments
-       * is already enough for the update
-       */
+
+      /* By default, the daemon performs change detection and can short-circuit
+       * if nothing has been appended/replace/deleted.
+       * However, when going through the editor flow, pre-filtering logic is applied
+       * client-side and the final command-line is directly passed to the daemon. */
```

Comment:
Might be worth linking to https://github.com/coreos/rpm-ostree/issues/2712 here.